app_test.go

```diff
@@ -2265,3 +2265,239 @@ func TestWhenExitSubCommandWithCodeThenAppQuitUnexpectedly(t *testing.T) {
+		t.Errorf("received error did not match expectation")
+	}
+}
+
+var duplicateCommandNamesTestsExpectedToError = map[string]struct {
+	args              []string
+	appName           string
+	commands          []Command
+	expectedErrString string
+}{
+	"duplicate command name": {
+		[]string{"say", "hello"},
+		"appname",
+		[]Command{
+			{
+				Name: "hello",
+				Action: func(c *Context) error {
+					fmt.Println("Hello")
+					return nil
+				},
+			},
+			{
+				Name: "hello",
+				Action: func(c *Context) error {
+					fmt.Println("Hello duplicate")
+					return nil
+				},
+			},
+		},
+		fmt.Sprintf("Your app contains multiple commands with the Name %q. Having multiple commands with the same name results in ambiguous behavior, so please make sure each command in your app has a unique name.", "hello"),
+	},
+	"duplicate subcommand name on command level": {
+		[]string{"say", "kind", "word"},
+		"appname",
+		[]Command{
+			{
+				Name: "kind",
+				Subcommands: []Command{
+					{
+						Name: "word",
+						Action: func(c *Context) error {
+							fmt.Println("please")
+							return nil
+						},
+					},
+					{
+						Name: "word",
+						Action: func(c *Context) error {
+							fmt.Println("thanks")
+							return nil
+						},
+					},
+				},
+			},
+		},
+		fmt.Sprintf("Your command %q contains multiple subcommands with the Name %q. Having multiple subcommands with the same name results in ambiguous behavior, so please make sure each subcommand in your command has a unique name.", "kind", "word"),
+	},
+	"duplicate subcommand name on subcommand level": {
+		[]string{"appname", "commandname", "subcommandname1", "subcommandname11"},
+		"appname",
+		[]Command{
+			{
+				Name: "commandname",
+				Subcommands: []Command{
+					{
+						Name: "subcommandname1",
+						Action: func(c *Context) error {
+							fmt.Println("this is subcommandname1")
+							return nil
+						},
+						Subcommands: []Command{
+							{
+								Name: "subcommandname11",
+								Action: func(c *Context) error {
+									fmt.Println("this is subcommandname11")
+									return nil
+								},
+							},
+							{
+								Name: "subcommandname11",
+								Action: func(c *Context) error {
+									fmt.Println("this is subcommandname11")
+									return nil
+								},
+							},
+						},
+					},
+					{
+						Name: "subcommandname2",
+						Action: func(c *Context) error {
+							fmt.Println("this is subcommandname2")
+							return nil
+						},
+					},
+				},
+			},
+		},
+		fmt.Sprintf("Your command %q contains multiple subcommands with the Name %q. Having multiple subcommands with the same name results in ambiguous behavior, so please make sure each subcommand in your command has a unique name.", "subcommandname1", "subcommandname11"),
+	},
+}
+
+func TestCheckDuplicateCommandNames(t *testing.T) {
```

Comment:
⭐⭐ _(seriously consider)_ These tests are all conceptually similar, consider moving the 4 test functions you have here (eg. `TestCheckDuplicateCommandNames`, `TestDuplicateSubcommandNameOnSubcommandLevel`, etc) into a single test function (eg. `TestCheckDuplicateCommandNames`)